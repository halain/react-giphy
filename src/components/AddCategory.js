import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ( {setCategories} ) => {
    
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) =>{
        setInputValue( e.target.value );
    }

    const handleSubmit = (e) =>{
        e.preventDefault();
        //console.log('submit');

        if (inputValue.trim().length>=2){
            //setCategories viene desde el componente padre. Se agrega el valor del input a las categorias mediante el operador ...
            setCategories( cat=>{
               return [inputValue, ...cat];
            });
            setInputValue('');
        }

    }
    
    return (
        <form onSubmit={handleSubmit}>
          <input 
                type="text"
                value={inputValue}  
                onChange={ handleInputChange }
           />  
        </form>
    )
}

AddCategory.protoTypes = {
    setCategories: PropTypes.func.isRequired
}