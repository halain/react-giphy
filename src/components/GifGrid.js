//import React, { useState, useEffect } from 'react';
import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
//import {getGifs} from '../helpers/getGifs'

export const GifGrid = ({category}) => {

   // const [images, setImages] = useState([]);
   

    //se dispara una unica vez cuando se renderiza el componente, asi si hay cambios en el componente no ejecuta constantemente la funcion con cada cambio
    //** */ useEffect( () =>{ funcion_a_ejecutar },[lista de dependencias]), si cambia la dependecia (category) se ejecuta nuevamente la funcion 
//     useEffect( () =>{
//         getGifs(category)
//             .then(setImages);
//    //         .then(imgs => setImages(imgs));
//     },[ category ]) 

   


    //** */ Crear custom Hook. Permitira que cuando el comp cargue, realizar la peticion fetch e indique cuando esta cargando y cuando termino
    //const state = useFetchGifs();
    //console.log(state);
    const {data:images, loading} = useFetchGifs(category);
    console.log(images, loading);

    return (
      <>
        <h3 className="animate__animated animate__fadeIn">{category}</h3>

        {loading && <p className="animate__animated animate__flash">Loading...</p>}

        <div className="card-grid">
          {images.map((img) => (
            <GifGridItem key={img.id} {...img} /> //...img envia cada propiedad de img en un objeto nuevo al componente GifGridItem
          ))}
        </div>

        
        
      </>
    );
}
