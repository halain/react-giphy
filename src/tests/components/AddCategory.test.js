import '@testing-library/jest-dom';
import React from 'react';
import { shallow } from 'enzyme';
import { AddCategory } from '../../components/AddCategory';

describe('pruebas en <AddCategory />', () => {

    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);

    beforeEach( ()=>{
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories} />);

    })

    test('debe renderizar correctamete el componente', () => {

        expect( wrapper ).toMatchSnapshot();
        
    });  
    
    
    test('debe de cambiar la caja de texto ', () => {
       
        const input = wrapper.find('input');
        const value = 'Hola mundo';
        input.simulate('change', { target: { value } });
         
        const p = wrapper.find('p');

        expect(p.text().trim()).toBe(value);

    });


    test('No debe postear la inofrmacion con submit ', () => {
        
        wrapper.find('form').simulate('submit', { preventDefault(){} } );

        expect( setCategories ).not.toHaveBeenCalled();

    });


    test('debe llamar el setCategorie y limpar la caja de texto ', () => {

        const input = wrapper.find('input');
        const value = 'Hola mundo';
        
        //simular input change
        input.simulate('change', { target: { value } });

        //simular submit
        wrapper.find('form').simulate('submit', { preventDefault(){} } );

        //setCategories debe ser llamado
        expect( setCategories ).toHaveBeenCalled(); //ha sido llamada
        expect( setCategories ).toHaveBeenCalledTimes(1); //llamada na sola ves
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) ); //llamado con una funcion como parametro

        //el valor del input esta en ''
        expect(input.prop('value')).toBe('');
        
    });
    
    
    
    
})
