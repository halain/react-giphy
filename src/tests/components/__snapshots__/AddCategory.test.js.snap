// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <AddCategory /> debe renderizar correctamete el componente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
